Adding devcontainer for easy Python development

storey247 opened the pull request:
Love the tool, wanted to contribute and here is the result :)

- Added a [VSCode devcontainer](https://code.visualstudio.com/docs/remote/containers) to help people get up and running with the repo quicker without needing to install any tooling or do any setup of venv
- Added a new rule to help when some repositories use `main` and some use `master` as the primary integration branch

Feedback welcome and thanks for the great tooling

scorphus commented on thefuck/rules/git_main_master.py:
```diff
@@ -0,0 +1,14 @@
+
+@git_support
+def match(command):
+    return "'master'" in command.output.lower() or "'main'" in command.output.lower()
```
Why is `.lower()` needed here?

scorphus commented on thefuck/rules/git_main_master.py:
```diff
@@ -0,0 +1,14 @@
+def get_new_command(command):
+    if "'master'" in command.output.lower():
+        return command.script.replace("master", "main")
+    else:
+        return command.script.replace("main", "master")
```
```suggestion
    return command.script.replace("main", "master")
```

scorphus reviewed (changes requested):
Looking good!

Would you be willing to separate the two features into two different commits and — if it's not too much to ask — two different PR's?

Thanks!

scorphus commented:
Also, the fix for Python 2.7 tests is already on current `master`, would you please rebase on top of that? 🙂

scorphus commented:
Howdy @storey247! Thanks for contributing. I'd really love to read your comments on my suggestions above. I understand that you're probably busy with many stuff. But I thought I'd ping you 🙂

storey247 commented:
hi @scorphus thanks for all the feedback, yes I will action and get this PR split into two :smile: apologies for the delayed response, work is a bit crazy atm

storey247 commented on .devcontainer/Dockerfile:
```diff
@@ -0,0 +1,22 @@
+ARG VARIANT="3"
+FROM mcr.microsoft.com/vscode/devcontainers/python:0-${VARIANT}
+
+# [Option] Install Node.js
+ARG INSTALL_NODE="true"
+ARG NODE_VERSION="lts/*"
+RUN if [ "${INSTALL_NODE}" = "true" ]; then su vscode -c "umask 0002 && . /usr/local/share/nvm/nvm.sh && nvm install ${NODE_VERSION} 2>&1"; fi
```
the devconainer.json file actually set the arg to false, so it's not getting installed, but as it's not necessary to have this I will remove 👍

storey247 replied:
I have also cleaned all this up as node is not required

storey247 commented on .devcontainer/Dockerfile:
```diff
@@ -0,0 +1,22 @@
+#     && apt-get -y install --no-install-recommends <your-package-list-here>
+
+# [Optional] Uncomment this line to install global node packages.
+# RUN su vscode -c "source /usr/local/share/nvm/nvm.sh && npm install -g <your-package-here>" 2>&1
```
will clean up

storey247 commented on .devcontainer/devcontainer.json:
```diff
@@ -0,0 +1,47 @@
+
+	// Comment out connect as root instead. More info: https://aka.ms/vscode-remote/containers/non-root.
+	//"remoteUser": "vscode"
+}
```
again, will clean

scorphus commented:
Awesome! I already split the commit into three, so no need to create another PR, I can merge the changes regarding the new rule and we leave this PR for the devcontainer addition. What do you think?

storey247 commented:
Amazing! Nice work! 👏👏👏

sounds like a plan, I’ll rebase this on Monday when the work is merged and I’ll tidy up the container so it’s nice and trim. 

Thanks @scorphus

scorphus commented:
There you go, @storey247. Sorry for the extra noise. To continue your work, I'd suggest you force-resetting to the `main_master` branch on your remote, that way you'll have exactly what this PR currently has. Please don't hesitate to ask me for help with that.

storey247 commented:
@scorphus I have now rebased my work from latest `master` and cleaned up the comments as discussed.

I also added some notes into the `Contribute.md` to help people get up and running with the devcontainer setup if required.

Hopefully now this work can be merged in too. Let me know if there is anything else you need me to look at

scorphus reviewed (approved):
Thats super, @storey247! Thanks so much for hanging in! Please consider my suggestion below.

storey247 commented on .devcontainer/devcontainer.json:
```diff
@@ -0,0 +1,38 @@
+		"context": "..",
+		"args": { 
+			// Update 'VARIANT' to pick a Python version: 3, 3.6, 3.7, 3.8, 3.9
+			"VARIANT": "3.9"
```
yes you are correct, the ARG will be overridden by the value passed in the `devcontainer.json` and acts like a fallback default. Will set the `VARIANT` to be 3 as suggested

storey247 replied:
done and sync'd to remote

storey247 commented:
> Thats super, @storey247! Thanks so much for hanging in! Please consider my suggestion below.

Feedback actions, just waiting on the builds and then will merge 👍

scorphus commented on .devcontainer/devcontainer.json:
```diff
@@ -0,0 +1,38 @@
+	"name": "Python 3",
+	"build": {
+		"dockerfile": "Dockerfile",
+		"context": "..",
+		"args": { 
+			// Update 'VARIANT' to pick a Python version: 3, 3.6, 3.7, 3.8, 3.9
+			"VARIANT": "3"
+		}
```
```suggestion
		"context": ".."
```
Let's remove that redundancy 🙂

scorphus commented on .devcontainer/devcontainer.json:
```diff
@@ -0,0 +1,38 @@
+	},
+
+	// Set *default* container specific settings.json values on container create.
+	"settings": { 
```
```suggestion
	"settings": {
```

scorphus commented on .devcontainer/devcontainer.json:
```diff
@@ -0,0 +1,38 @@
+	],
+
+	// Use 'postCreateCommand' to run commands after the container is created.
+	"postCreateCommand": "pip3 install -r requirements.txt && python3 setup.py develop",
```
```suggestion
	"postCreateCommand": "pip3 install -r requirements.txt && python3 setup.py develop"
```

scorphus reviewed (approved):
Thanks, @storey247! That's a great contribution!

Also, thanks for the new `git_main_master` rule! 🙌
